internal/provider/deployment.go

```diff
@@ -79,68 +80,72 @@ func getDeploymentDeployer(d *schema.ResourceData, apiClient *apiClient) (Deploy
+	deploymentDataBytes, err := json.Marshal(deploymentData)
 	if err != nil {
-		log.Printf("error parsing deploymentdata: %s", err.Error())
+		log.Printf("error parsing deployment data: %s", err.Error())
 	}
 
-	networkingState := apiClient.state.GetNetworkState()
+	networkingState := threefoldPluginClient.state.GetState().Networks
 	net := networkingState.GetNetwork(networkName)
 	ipRange := net.GetNodeSubnet(nodeID)
 
 	deploymentDeployer := DeploymentDeployer{
-		Id:          d.Id(),
-		Node:        nodeID,
-		Disks:       disks,
-		VMs:         vms,
-		QSFSs:       qsfs,
-		ZDBs:        zdbs,
-		IPRange:     ipRange,
-		NetworkName: networkName,
-		APIClient:   apiClient,
-		ncPool:      pool,
-		deployer:    deployer.NewDeployer(apiClient.identity, apiClient.twin_id, apiClient.grid_client, pool, true, solutionProvider, string(deploymentDataStr)),
+		ID:                    d.Id(),
+		Node:                  nodeID,
+		Disks:                 disks,
+		VMs:                   vms,
+		QSFSs:                 qsfs,
+		ZDBs:                  zdbs,
+		IPRange:               ipRange,
+		NetworkName:           networkName,
+		ThreefoldPluginClient: threefoldPluginClient,
+		ncPool:                pool,
+		deployer:              deployer.NewDeployer(threefoldPluginClient.identity, threefoldPluginClient.twinID, threefoldPluginClient.gridProxyClient, pool, true, solutionProvider, string(deploymentDataBytes)),
 	}
 	return deploymentDeployer, nil
 }
 
 func (d *DeploymentDeployer) assignNodesIPs() error {
-	networkingState := d.APIClient.state.GetNetworkState()
+	networkingState := d.ThreefoldPluginClient.state.GetState().Networks
 	network := networkingState.GetNetwork(d.NetworkName)
-	usedIPs := network.GetNodeIPsList(d.Node)
+	usedHosts := network.GetUsedNetworkHostIDs(d.Node)
 	if len(d.VMs) == 0 {
 		return nil
 	}
-	_, cidr, err := net.ParseCIDR(d.IPRange)
+	ip, ipRangeCIDR, err := net.ParseCIDR(d.IPRange)
 	if err != nil {
 		return errors.Wrapf(err, "invalid ip %s", d.IPRange)
 	}
 	for _, vm := range d.VMs {
-		if vm.IP != "" && cidr.Contains(net.ParseIP(vm.IP)) && !Contains(usedIPs, net.ParseIP(vm.IP)[3]) {
-			usedIPs = append(usedIPs, net.ParseIP(vm.IP)[3])
+		vmIP := net.ParseIP(vm.IP)
+		vmHostID := vmIP[3]
+		if vm.IP != "" && ipRangeCIDR.Contains(vmIP) && !Contains(usedHosts, vmHostID) {
+			usedHosts = append(usedHosts, vmHostID)
 		}
 	}
-	cur := byte(2)
+	curHostID := byte(2)
+
 	for idx, vm := range d.VMs {
-		if vm.IP != "" && cidr.Contains(net.ParseIP(vm.IP)) {
+
+		if vm.IP != "" && ipRangeCIDR.Contains(net.ParseIP(vm.IP)) {
 			continue
 		}
-		ip := cidr.IP
-		ip[3] = cur
-		for Contains(usedIPs, ip[3]) {
-			if cur == 254 {
+
+		for Contains(usedHosts, curHostID) {
+			if curHostID == 254 {
 				return errors.New("all 253 ips of the network are exhausted")
 			}
-			cur++
-			ip[3] = cur
+			curHostID++
 		}
-		d.VMs[idx].IP = ip.String()
-		usedIPs = append(usedIPs, ip[3])
+		usedHosts = append(usedHosts, curHostID)
+		vmIP := ip
```

Comment:
i had a problem with deploying a vm because it had an ipv6, which was not in the subnet assigned to this node, i guess the ip returned from net.ParseCIDR() defaults to ipv6?
anyway, changing this to ip.To4() solved it.